pkg/block/fetcher.go

```diff
@@ -298,6 +298,14 @@ func (s *MetaFetcher) Fetch(ctx context.Context) (metas map[ulid.ULID]*metadata.
 			return nil
 		}
 
+		deletionMark, err := metadata.ReadDeletionMark(ctx, s.bkt, s.logger, id.String())
+		if err != nil && errors.Cause(err) != metadata.ErrorDeletionMarkNotFound {
+			return err
+		}
+		if deletionMark != nil {
+			return nil
+		}
+
```

Comment:
while we are fetching metas, we need to remove the blocks which are scheduled to be deleted. Otherwise it will consider the blocks which are marked to be deleted.
This makes sure that calling `block.Delete` has the same effect as `block.MarkForDeletion`. 
This is one way to do it.

Another way would be to have a filter that filters out blocks that have `deletion-mark.json` file, but since we only get meta in Filter, we don't get `deletion-mark.json` file. We can store bkt in the struct of the filter, so in a way the above code will move to the filter.

Should we add this code in another filter or here is fine? Please comment @bwplotka @pstibrany @GiedriusS

Reply:
Having this in metafetcher is good because we should check the `deletion-mark.json` file every time we call fetch. This is to make sure that `block.MarkForDeletion` has same effect as `block.Delete`. Otherwise even if we call MarkForDeletion, the block will be still fetched.

Having it as a filter makes it optional to have, and having it in metafetcher makes it mandatory.
I am for keeping it mandatory, let me know your thoughts 🤗

Reply:
> Another way would be to have a filter that filters out blocks that have deletion-mark.json 

Hm, maybe it's confusing but we already have dedup filter so we won't touch those, right?

Reply:
I mean since it's faster, we could have just filter for this even for reader yes. HOWEVER for readers (and compactor as well) we need to make sure the replacement block is *there*

So I think deletion mark might actually have optional field for the ID of block that you should check first if you fully see it, otherwise ignore this block, right? (:

Reply:
Let's move this question to main thread, otherwise it's lost.